src/compute.rs

```
@@ -8,54 +9,65 @@ use crate::{
 
-    hasher.update(prefix.as_bytes());
-    hasher.update(id.to_bytes());
     for b in B {
-        hasher.update(b.D.compress().as_bytes());
-        hasher.update(b.E.compress().as_bytes());
+        buf.extend_from_slice(b.D.compress().as_bytes());
+        buf.extend_from_slice(b.E.compress().as_bytes());
     }
-    hasher.update(msg);
 
-    hash_to_scalar(&mut hasher)
+    buf.extend_from_slice(msg);
+
+    expand_to_scalar(&buf, prefix)
 }
 
 #[allow(non_snake_case)]
-/// Compute a binding value from the party ID, public nonces, and signed message
-pub fn binding_compressed(id: &Scalar, B: &[(Compressed, Compressed)], msg: &[u8]) -> Scalar {
-    let mut hasher = Sha256::new();
-    let prefix = "WSTS/binding";
+/// Compute a binding value from the party ID, public nonces, and signed message using XMD-based expansion.
+pub fn binding_compressed(
+    id: &Scalar,
+    B: &[(Compressed, Compressed)],
+    msg: &[u8],
+) -> Result<Scalar, EllipticCurveError> {
+    let prefix = b"WSTS/binding";
+
+    // Serialize all input into a buffer
+    let mut buf = Vec::new();
+    buf.extend_from_slice(&id.to_bytes());
 
-    hasher.update(prefix.as_bytes());
-    hasher.update(id.to_bytes());
     for (D, E) in B {
-        hasher.update(D.as_bytes());
-        hasher.update(E.as_bytes());
+        buf.extend_from_slice(D.as_bytes());
+        buf.extend_from_slice(E.as_bytes());
     }
-    hasher.update(msg);
 
-    hash_to_scalar(&mut hasher)
+    buf.extend_from_slice(msg);
+
+    expand_to_scalar(&buf, prefix)
 }
 
 #[allow(non_snake_case)]
-/// Compute the schnorr challenge from the public key, aggregated commitments, and the signed message
-pub fn challenge(publicKey: &Point, R: &Point, msg: &[u8]) -> Scalar {
+/// Compute the schnorr challenge from the public key, aggregated commitments, and the signed message using XMD-based expansion.
+pub fn challenge(publicKey: &Point, R: &Point, msg: &[u8]) -> Result<Scalar, EllipticCurveError> {
```

This function should return a WSTS error, not a foreign one.

---

Yeah, maybe we don't need the error here at all.

---

Done in 776c669eaaa6eb030db02e1903dcc2779095fa96